pkg/controller/daemon/update_test.go

```diff
@@ -294,6 +298,219 @@ func TestDaemonSetUpdatesSomeOldPodsNotReady(t *testing.T) {
+		t.Fatal(err)
+	}
+	expectSyncDaemonSets(t, manager, ds, podControl, 100, 0, 0)
+	markPodsReady(podControl.podStore)
+	var hash1 string
+	// at this point we have 100 pods running from daemonset, and we mark
+	// the controller has which will be used later on to fake old pods
+	for _, obj := range podControl.podStore.List() {
+		pod := obj.(*v1.Pod)
+		hash1 = pod.Labels[apps.ControllerRevisionHashLabelKey]
+		break
+	}
+
+	ds.Spec.Template.Spec.Containers[0].Image = "foo2/bar2"
+	ds.Spec.UpdateStrategy.Type = apps.RollingUpdateDaemonSetStrategyType
+	maxUnavailable := 10
+	intStr := intstr.FromInt(maxUnavailable)
+	ds.Spec.UpdateStrategy.RollingUpdate = &apps.RollingUpdateDaemonSet{MaxUnavailable: &intStr}
+	err = manager.dsStore.Update(ds)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// we need to iterate 10 times, since we allow 10 max unavailable, to reach 100 nodes rollout
+	for i := 0; i < 10; i++ {
+		clearExpectations(t, manager, ds, podControl)
+		expectSyncDaemonSets(t, manager, ds, podControl, 0, maxUnavailable, 0)
+
+		clearExpectations(t, manager, ds, podControl)
+		expectSyncDaemonSets(t, manager, ds, podControl, maxUnavailable, 0, 0)
+		// make sure to mark the pods ready, otherwise the followup rollouts will fail
+		markPodsReady(podControl.podStore)
+	}
+
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
+	clearExpectations(t, manager, ds, podControl)
+
+	// to reach the following situation
+	// - maxUnavailable 10
+	// - 88 unavailable new pods
+	// - 2 unavailable old pods
+	// - 10 available old pods
+	oldUnavailablePods := []string{}
+	for i, obj := range podControl.podStore.List() {
+		pod := obj.(*v1.Pod)
+		// mark the latter 90 pods not ready
+		if i >= 10 {
+			condition := v1.PodCondition{Type: v1.PodReady, Status: v1.ConditionFalse}
+			podutil.UpdatePodCondition(&pod.Status, &condition)
+		}
+		// mark the first 12 pods with older hash
+		if i < 12 {
+			pod.Labels[apps.ControllerRevisionHashLabelKey] = hash1
+			// note down 2 not available old pods
+			if i >= 10 {
+				oldUnavailablePods = append(oldUnavailablePods, pod.Name)
+			}
+		}
+	}
+
+	clearExpectations(t, manager, ds, podControl)
+	t.Logf("expect 10 old pods deletion in 1st iteration")
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 2, 0)
+
+	clearExpectations(t, manager, ds, podControl)
+	t.Logf("expect 10 new pods creation in 2nd iteration")
+	expectSyncDaemonSets(t, manager, ds, podControl, 2, 0, 0)
+
+	clearExpectations(t, manager, ds, podControl)
+	t.Logf("expect no modifications in 3rd iteration")
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
+	clearExpectations(t, manager, ds, podControl)
+
+	// check if oldUnavailablePods were replaced
+	t.Logf("Looking for old pods %s", strings.Join(oldUnavailablePods, ", "))
+	notUpdatedOldPods := []string{}
+	for _, obj := range podControl.podStore.List() {
+		pod := obj.(*v1.Pod)
+		for _, oldPod := range oldUnavailablePods {
```

Comment:
`oldUnavailablePods` would be better as a set

Reply:
Done

Reply:
This one still applies. We can either use `Has` here, or do an intersection later.

Reply:
Sorry, missed it.